### ISSUE TITLE
Some web application keybindings are captured by Nyxt instead of the web-buffer in `vi-insert-mode`

### ISSUE THREAD

**Issue by Kabouik** [atlas-engineer/nyxt#1635]
**Describe the bug**
When on a webpage that supports keybindings, some are not properly captured by the web buffer, even when in `vi-insert-mode` (I would **not** expect them to be captured when in `vi-normal-mode`). 

For instance, with the mattermost web client, `<C-i>` or `<C-n>` which should be formatting keybindings instead call the Autofill and Bookmark functions in Nyxt, despite `vi-insert-mode` being active. `<C-u>` should open the "upload file" dialog, but does nothing (it did work once when I pressed it twice in a row, but I couldn't reproduce it). 

Some of them are captured however, immediately like `/` to focus the search field on Github, or when pressed twice like `<C-u>` on Discord to toggle the user list.

Those keybindings work fine in Epiphany. I have not tried with CUA or Emacs mode but wouldn't be surprised if this was a consequence of the vi modal system.

I understand that this is a difficult thing to implement without compromises. For instance, I really like that I can still switch buffers with `<C-[>` and `<C-]>` without opening the prompt-buffer or without leaving `vi-insert-mode` when I am composing, and I am not sure how this could still work if all keybindings were interpreted in priority by the web-buffer when in insert mode. Similarly, it would become impossible to call the promp buffer when in `vi-insert-mode`, while this can be done currently with `<C-space>`; one would have to hit `<Esc>` and then `:` or `<C-space>`.

**Information**
- OS name+version: Solus (rolling), Debian Sid (rolling)
- Graphics card and driver: Intel, Nvidia MX150 with nvidia-glx-driver, or none
- Desktop environment / Window manager name+version: i3wm
- How you installed Nyxt (Guix pack, package manager, build from source): built from `master`
- Information from command copy-system-information:

```
Nyxt version: 2.1.1-227-g85bbe1c2
Renderer version: GI-GTK
Operating system kernel: Linux 5.13.2-188.current
Lisp implementation: SBCL 2.1.2 (Dynamic space size: 1073741824)
Features: (WEBKIT2 WEBKIT2-2.32 WEBKIT2-PASTE-PLAINTEXT WEBKIT2-TRACKING WEBKIT2-MUTE
 WEBKIT2-EMOJI WEBKIT2-MEDIA WEBKIT2-SANDBOXING GTK-3-22 GTK-3-20 GTK-3-18
 GTK-3-16 GTK-3-14 GTK-3-12 GTK-3-10 GTK-3-8 GTK-3-6 GTK-3-4 GTK GDK-3-22
 GDK-3-20 GDK-3-18 GDK-3-16 GDK-3-14 GDK-3-12 GDK-3-10 GDK-3-8 GDK-3-6 GDK-3-4
 CAIRO-1-10 CAIRO-1-12 GDK-PIXBUF GLIB-2-30 GLIB-2-32 GLIB-2-34 GLIB-2-36
 GLIB-2-38 GLIB-2-40 GLIB-2-42 GLIB-2-44 GLIB-2-46 GLIB-2-48 GLIB-2-50
 GLIB-2-52 GLIB-2-54 GLIB-2-56 GLIB-2-58 GLIB NYXT-2 FSET-EXT-STRINGS
 CUSTOM-HASH-TABLE-NATIVE SWANK PLUMP-UTF-32 GLOBAL-VARS DECLARE-TYPES
 PARENSCRIPT NAMED-READTABLES 21BIT-CHARS CHUNGA FLEXI-STREAMS CLOSER-MOP
 CL-PPCRE-UNICODE CL-UNICODE CL-PPCRE BORDEAUX-THREADS SPLIT-SEQUENCE
 GRAY-STREAMS FAST-IO-SV FAST-IO SBCL-USES-SB-ROTATE-BYTE
 ASDF-SYSTEM-CONNECTIONS CL-JSON-CLOS CL-JSON THREAD-SUPPORT FLAT-NAMESPACE
 X86-64 UNIX CFFI FLAT-NAMESPACE SEQUENCE-EMPTYP QUICKLISP ASDF3.3 ASDF3.2
 ASDF3.1 ASDF3 ASDF2 ASDF OS-UNIX NON-BASE-CHARS-EXIST-P ASDF-UNICODE X86-64
 GENCGC 64-BIT ANSI-CL COMMON-LISP ELF IEEE-FLOATING-POINT LINUX LITTLE-ENDIAN
 PACKAGE-LOCAL-NICKNAMES SB-CORE-COMPRESSION SB-LDB SB-PACKAGE-LOCKS SB-THREAD
 SB-UNICODE SBCL UNIX)

ASDF version: 3.3.1
Critical dependencies: (/home/mathieu/Projects/common-lisp/nyxt/_build/submodules/cl-cffi-gtk/gtk/cl-cffi-gtk.asd
 /home/mathieu/Projects/common-lisp/nyxt/_build/quicklisp-client/dists/quicklisp/software/cl-gobject-introspection-20210124-git/cl-gobject-introspection.asd
 /home/mathieu/Projects/common-lisp/nyxt/_build/submodules/cl-webkit/webkit2/cl-webkit2.asd)

Quicklisp dist version: 2021-06-30
Quicklisp client version: 2020-01-04
Local project directories: (/home/mathieu/Projects/common-lisp/nyxt/_build/submodules/
 /home/mathieu/Projects/common-lisp/nyxt/_build/quicklisp-client/local-projects/)
Critical dependencies(#<SYSTEM cl-cffi-gtk / cl-cffi-gtk-20201220-git / quicklisp 2021-06-30>
 #<SYSTEM cl-gobject-introspection / cl-gobject-introspection-20210124-git / quicklisp 2021-06-30>
 #<SYSTEM cl-webkit2 / cl-webkit-20210630-git / quicklisp 2021-06-30>)
```

**Comment by jmercouris** [comment]
This is because of the way keybindings coalesce. If something isn't found in a most specific keybinding, Nyxt will look in a more generalized keybinding. In order to work around this you can invoke `application-mode` which will pass through all keybindings to the web application.

**Comment by jmercouris** [comment]
Perhaps we should call it `passthrough-mode` or something, because it seems few people find it!

**Comment by Kabouik** [comment]
I missed it indeed (but I would probably have missed `passthrough-mode` just equally), I didn't have the reflex to look for modes.

That doesn't seem to work very well with `vi` modes though. If I go into `vi-insert-mode` and then enable `application-mode`, then `<C-b>` and `<C-i>` do indeed trigger the markdown formatting characters in mattermost, but `<C-u>` still doesn't invoke the "upload file" dialog. Additionally, if I enable the mode when in `vi-normal-mode`, then obviously I cannot go into `vi-insert-mode` because `i` is passed through to the web app; but I'm not sure how this could be worked around without breaking functionality.

Another issue is once you are in `application-mode`, since no keybindings are interpreted by Nyxt, you can no longer disabled it with the keyboard because you cannot open the prompt with key bindings or even exit the `vi-insert-mode`. The only way I found was to use the mouse to click on the "Enable modes" indicator in the toolbar to show the `toggle-modes` prompt, and disable it from there.

**Comment by Ambrevar** [comment]
It's a bug that I've introduced recently, I was planning on working on it last week, will get to it as soon as possible.

**Comment by Ambrevar** [comment]
My bad, I thought it was about something else, and after some investigation the bug I should I had introduced is not one.

Allow to clarify a few things:

- vi-normal-mode does not forward anything to the web renderer.
- vi-normal-mode inherits from 2 scheme: emacs and CUA.
- vi-insert-mode inherits from 1 scheme: CUA.
  This is why `C-i` does not work in vi-insert-mode: it's bound in CUA mode.
- The override-map in which C-space is bound can never be bypassed, so even if you enable application-mode, `C-space` will still work.

A few options:

- We can remove the CUA inheritance in `vi-insert-mode`, but that would come at a cost, as @Kabouik mentioned: the user won't be able to press familiar shortcuts.
- We can enable application-mode when entering `vi-insert-mode`.  We could add a binding to toggle it, e.g. `C-z`.

@Kabouik Why doesn't `C-u` get forwarded, any idea?

**Comment by Ambrevar** [comment]
I just tried `C-u` on Mattermost, works for me.

So I'm in favour of adding the following:

- An `application-mode` toggler binding
- An `application-mode` escape binding (just for the next key sequence)
- An vi-mode option to choose whether application-mode should be enabled when switching to insert mode.

**Comment by Kabouik** [comment]
Thanks for looking into it @Ambrevar. That's weird, I checked again `C-u` in Mattermost, and like I said in a previous message, it worked once with `application-mode` and `vi-insert-mode` enabled (after pressing `C-u` twice, for some reason) but then didn't work again after I escaped the file-upload dialog that one time. Then, doesn't matter how many times I pressed `C-u` or which modes were enabled, the file-dialog prompt wouldn't show up. My Mattermost server is not on the latest version, but `C-u` is the correct binding for uploading files.

```
<INFO> [11:52:50] Pressed keys: g
<INFO> [11:52:52] Loading "https://mattermost.[redacted].fr/lamouze/channels/town-square".
<INFO> [11:52:54] Finished loading "https://mattermost.[redacted].fr/lamouze/channels/town-square".
<INFO> [11:53:53] Pressed keys: C-u
<INFO> [11:53:56] Pressed keys: C-u
<INFO> [11:53:57] Pressed keys: C-u
<INFO> [11:53:58] Pressed keys: C-u
<WARN> [11:54:00] Warning: Error on separate thread: Condition NYXT::NYXT-PROMPT-BUFFER-CANCELED was signalled.
<INFO> [11:54:03] Pressed keys: C-u
<INFO> [11:54:06] Pressed keys: C-u
<INFO> [11:54:08] Pressed keys: C-u
<INFO> [11:54:09] Pressed keys: C-u
<INFO> [11:54:09] Pressed keys: C-u
<INFO> [11:54:10] Pressed keys: C-u
<INFO> [11:54:13] Pressed keys: C-u
<INFO> [11:54:15] Pressed keys: C-u
<INFO> [11:54:16] Pressed keys: C-u
<INFO> [11:54:17] Pressed keys: C-u
<INFO> [11:54:19] Pressed keys: C-u
<INFO> [11:54:20] Pressed keys: C-u
<INFO> [11:54:21] Pressed keys: C-u
<INFO> [11:54:21] Pressed keys: C-u
<INFO> [11:54:22] Pressed keys: C-u
<INFO> [11:54:23] Pressed keys: C-u
<INFO> [11:54:26] Pressed keys: g
<INFO> [11:54:30] Loading "https://github.com/atlas-engineer/nyxt/issues/1636".
<INFO> [11:54:33] Finished loading "https://github.com/atlas-engineer/nyxt/issues/1636".
```

I can confirm also that `C-space` has no effect on my end when `application-mode` is enabled, so I have to use my mouse to click on the Enabled modes to show the `toggle-mode` list and disable it from there. I'm on `master`. I pressed `C-space` before loading Github in the above history, and yet it doesn't appear in the logs.

I get the same results when running `nyxt -IC`, except for the `C-space` issue, so that one is apparently definitely related to my config. Not sure where is the issue yet, but it's probably not a bug.

**Comment by Ambrevar** [comment]
> Thanks for looking into it @Ambrevar. That's weird, I checked again `C-u` in Mattermost, and like I said in a previous message, it worked once with `application-mode` and `vi-insert-mode` enabled (after pressing `C-u` twice, for some reason) but then didn't work again after I escaped the file-upload dialog that one time. Then, doesn't matter how many times I pressed `C-u` or which modes were enabled, the file-dialog prompt wouldn't show up. My Mattermost server is not on the latest version, but `C-u` is the correct binding for uploading files.

This is a known bug, but it's unrelated to VI / key bindings.

@aartaka run-file-chooser (and possibly script-dialog) works only once.
Any idea what that is?

> I get the same results when running `nyxt -IC`, except for the `C-space` issue, so that one is apparently definitely related to my config. Not sure where is the issue yet, but it's probably not a bug.

Indeed, can you share your config?  Maybe we are not smart enough on
Nyxt end.

**Comment by Kabouik** [comment]
> Indeed, can you share your config? Maybe we are not smart enough on

Sure, you can find it [here](https://git.teknik.io/matf/nyxt/src/branch/master/.config/nyxt).

**Comment by aartaka** [comment]
> @aartaka run-file-chooser (and possibly script-dialog) works only once. Any idea what that is?

That's a bug! :D 

I'll investigate. I'm most probably not catching some prompt-buffer conditions or not freeing some WebKit resources.

Script-dialogs work file, I've just tested. It's only file chooser.

**Comment by aartaka** [comment]
17d6a5a9666991ef45fbc8fd5811c5d68e095c47 should do the trick.

**Comment by Ambrevar** [comment]
Thanks @aartaka !

@Kabouik I've sent a pull request:  #1652.
Let me know what you think :)

**Comment by Kabouik** [comment]
It's a rip off from @ericdrgn's configuration with just easy customizations, all the hard work is from him and others on IRC!

Thank you @aartaka and @Ambrevar for the fixes, I'll try upgrading a little bit later and will let you know if there are still issues on my end.

**Comment by Ambrevar** [comment]
@Kabouik, I just had a look at your config and could spot two issues:

- urlprompt `*my-search-engines*` lacks closing parenthesis.

- `C-space` is indeed unbound when you replace the `override-map`.  But
  this should be OK because you bind `M-x` there to `execute-command`.

There are also a few warnings about unused arguments, you might want to
check that out.

Hope that helps ;)

**Comment by Kabouik** [comment]
Awesome, thanks for taking the time!

- I fixed `urlprompt.lisp`

- Indeed, `M-x` does work to show the prompt even when `application-mode` is enabled. This will only be an issue on my other device where I can't use Meta, but I should think about an alternative keybinding for essential commands like that one. Is there any technical reason for disabling `C-space` when `override-map` is set, even though `C-space` is not defined into it?

- Yes I saw those warnings but procrastinated so far because those are commands I didn't write myself (just like most of the config actually!) and didn't know from the top of my head how to fix those, plus the associated config is not central to my Nyxt use.

**Comment by Ambrevar** [comment]
>  Is there any technical reason for disabling `C-space` when `override-map` is set, even though `C-space` is not defined into it?

Not sure what you mean.
The default override-map is 

```lisp
(let ((map (make-keymap "override-map")))
                   (define-key map
                     "C-space" 'execute-command))
```

so if you replace the override map with one that does not bind `C-space`, then it won't be bound.  Does that make sense?

**Comment by Kabouik** [comment]
It totally makes sense. I got confused because `C-space` still invokes the prompt buffer with my current configuration despite not being manually set, so I thought it was default, but in fact it's probably inherited from somewhere else and is inhibited when `application-mode` is disabled.

Now that I understand better, I just added `C-space` and `:` as keybindings to invoke the prompt buffer, so that they can override `application-mode` in all cases.

**Comment by Kabouik** [comment]
Manually setting `:` in my `override-map` was not such a great idea because it also overrides `vi-input-mode`, meaning you can't type `:` in a text field anymore. It's a bit unexpected because I think `:` is default for `execute-command` in vi bindings, so I would assume it should still invoke the prompt when both `application-mode` and `vi-normal-mode` are enabled, and just type the `:` character whenever `vi-insert-mode` is enabled, regardless of `application-mode`.

Yet the only way to invoke the prompt buffer with `:` when `application-mode` is enabled is to set `:` in my `override-map`, with the bad side effect described above. Not important anyway, I'm happy with `C-space` and `M-x` to get into the prompt when `application-mode` is enabled.

**Comment by Ambrevar** [comment]
Indeed, `override-map` overrides everything, including character insertion! :p 

So it only makes sense to include modifier-prefixed bindings in the override-map.

Do you think the documentation should be clearer about this?

>  so I would assume it should still invoke the prompt when both `application-mode` and `vi-normal-mode`

Application-mode has priority over vi-normal-mode, so it's expected that `:` won't trigger the vi-normal-mode, otherwise how could we pass `:` to the web view?

As I understand the problem, there is no way around this binding pickle.  Can you think of something?

**Comment by Kabouik** [comment]
I don't feel there is a huge need for it, but if it does not hurt, then definitely adding to the documentation that "`override-map` takes priority over everything, including text insertion, and is therefore better used with modifier-prefixed bindings" might be useful. Speaking of it, what I think is a bigger issue is that the prompt-buffer does not show custom bindings defined in `override-map`, just default ones.

Yes I agree that there would be no way to pass `:` to the web view if `application-mode` has priority, but perhaps user would want to decide what should have priority depending on the bindings they need in their web applications. This way they dould set vi-modes at a higher rank than `application-mode` and use `i`/`Esc`/`:` to manipulate vi-modes and the prompt even when `application-mode` is on, text insertion when `vi-insert-mode` is enabled together with `application-mode`, and web app bindings when `vi-normal-mode` is set on top of `application-mode`. Or would it be broken too?

**Comment by Kabouik** [comment]
I can confirm that `C-u` now works multiple times in Mattermost too. Thanks @aartaka and @Ambrevar for the fast PRs!

**Comment by aartaka** [comment]
> Yes I agree that there would be no way to pass `:` to the web view since `application-mode` has priority, but perhaps users would want to decide what should have priority depending on the bindings they need in their web applications. This way they could set vi-modes at a higher rank than `application-mode` and use `i`/`Esc`/`:` to manipulate vi-modes and the prompt even when `application-mode` is on, text insertion when `vi-insert-mode` is enabled together with `application-mode`, and web app bindings when `vi-normal-mode` is set on top of `application-mode`. Or would it be broken too?

The purpose of `application-mode` is exactly to disable any Nyxt keybindings, especially those from `vi-normal-mode`.

**Comment by Ambrevar** [comment]
I've updated the doc in c4f0b78b7120eb60960b6181ec0192cae755a90b.

**Comment by Ambrevar** [comment]
What you are suggestion with a vi-mode override is doable through the
override map: you can customize a mode to add a binding to the override
map.  This binding could then handle `:`, `i` and `Esc` the way you
want.

**Comment by Ambrevar** [comment]
It seems that this is all addressed, isn't it?  Shall we close the issue?

**Comment by aartaka** [comment]
Yes, sounds complete.
